File: src/test/java/org/springframework/samples/petclinic/rest/contract/OwnerApiContractTests.java
package org.springframework.samples.petclinic.rest.contract;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.resttestclient.TestRestTemplate;
import org.springframework.boot.resttestclient.autoconfigure.AutoConfigureTestRestTemplate;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.samples.petclinic.rest.dto.OwnerCreateDto;
import org.springframework.samples.petclinic.rest.dto.OwnerDto;
import org.springframework.samples.petclinic.rest.dto.OwnerUpdateDto;
import org.springframework.samples.petclinic.rest.dto.PagedResultDto;
import org.springframework.test.annotation.DirtiesContext;

@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@AutoConfigureTestRestTemplate
class OwnerApiContractTests {

	@Autowired
	private TestRestTemplate restTemplate;

	@Test
	void listOwnersReturnsPagedResult() {
		ResponseEntity<PagedResultDto> response = restTemplate.getForEntity("/api/v1/owners", PagedResultDto.class);
		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
		assertThat(response.getBody()).isNotNull();
		assertThat(response.getBody().content()).isNotNull();
		assertThat(response.getBody().pageNumber()).isGreaterThan(0);
		assertThat(response.getBody().pageSize()).isGreaterThan(0);
		assertThat(response.getBody().totalElements()).isGreaterThan(0);
		assertThat(response.getBody().totalPages()).isGreaterThan(0);
	}

	@Test
	void listOwnersSupportsPagination() {
		ResponseEntity<PagedResultDto> page1 = restTemplate.getForEntity("/api/v1/owners?page=1&size=2",
				PagedResultDto.class);
		assertThat(page1.getBody()).isNotNull();
		assertThat(page1.getBody().content().size()).isLessThanOrEqualTo(2);

		ResponseEntity<PagedResultDto> page2 = restTemplate.getForEntity("/api/v1/owners?page=2&size=2",
				PagedResultDto.class);
		assertThat(page2.getBody()).isNotNull();
	}

	@Test
	void getOwnerReturnsCompleteDto() {
		ResponseEntity<OwnerDto> response = restTemplate.getForEntity("/api/v1/owners/1", OwnerDto.class);
		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
		OwnerDto owner = response.getBody();
		assertThat(owner).isNotNull();
		assertThat(owner.id()).isNotNull();
		assertThat(owner.firstName()).isNotNull();
		assertThat(owner.lastName()).isNotNull();
		assertThat(owner.address()).isNotNull();
		assertThat(owner.city()).isNotNull();
		assertThat(owner.telephone()).isNotNull();
	}

	@Test
	void createOwnerReturnsCreatedWithLocation() {
		OwnerCreateDto dto = new OwnerCreateDto("Jane", "Doe", "456 Oak Ave", "Portland", "5035551234");
		ResponseEntity<OwnerDto> response = restTemplate.postForEntity("/api/v1/owners", dto, OwnerDto.class);
		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
		assertThat(response.getBody()).isNotNull();
		assertThat(response.getBody().id()).isNotNull();
		assertThat(response.getBody().firstName()).isEqualTo("Jane");
		assertThat(response.getBody().lastName()).isEqualTo("Doe");
	}

	@Test
	void createOwnerValidatesAllFields() {
		OwnerCreateDto blankDto = new OwnerCreateDto("", "", "", "", "short");
		ResponseEntity<String> response = restTemplate.postForEntity("/api/v1/owners", blankDto, String.class);
		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
	}

	@Test
	void updateOwnerModifiesExistingOwner() {
		OwnerUpdateDto dto = new OwnerUpdateDto("UpdatedName", "Franklin", "110 W. Liberty St.", "Madison",
				"6085551023");
		ResponseEntity<OwnerDto> response = restTemplate.exchange("/api/v1/owners/1", HttpMethod.PUT,
				new HttpEntity<>(dto), OwnerDto.class);
		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
		assertThat(response.getBody().firstName()).isEqualTo("UpdatedName");
	}

	@Test
	void deleteOwnerRemovesOwner() {
		OwnerCreateDto dto = new OwnerCreateDto("Delete", "Owner", "789 Pine", "Town", "9998887776");
		ResponseEntity<OwnerDto> createResponse = restTemplate.postForEntity("/api/v1/owners", dto, OwnerDto.class);
		Integer id = createResponse.getBody().id();

		ResponseEntity<Void> deleteResponse = restTemplate.exchange("/api/v1/owners/" + id, HttpMethod.DELETE, null,
				Void.class);
		assertThat(deleteResponse.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);

		ResponseEntity<String> getResponse = restTemplate.getForEntity("/api/v1/owners/" + id, String.class);
		assertThat(getResponse.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
	}

}

File: src/test/kotlin/org/springframework/samples/petclinic/rest/contract/OwnerApiContractTests.kt
package org.springframework.samples.petclinic.rest.contract

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.resttestclient.TestRestTemplate
import org.springframework.boot.resttestclient.autoconfigure.AutoConfigureTestRestTemplate
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.http.HttpEntity
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.samples.petclinic.rest.dto.OwnerCreateDto
import org.springframework.samples.petclinic.rest.dto.OwnerDto
import org.springframework.samples.petclinic.rest.dto.OwnerUpdateDto
import org.springframework.samples.petclinic.rest.dto.PagedResultDto
import org.springframework.test.annotation.DirtiesContext

@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@AutoConfigureTestRestTemplate
class OwnerApiContractTests {

    @Autowired
    private lateinit var restTemplate: TestRestTemplate

    @Test
    fun listOwnersReturnsPagedResult() {
        val response: ResponseEntity<PagedResultDto<Any>> = restTemplate.getForEntity("/api/v1/owners", PagedResultDto::class.java) as ResponseEntity<PagedResultDto<Any>>
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body).isNotNull()
        assertThat(response.body!!.content).isNotNull()
        assertThat(response.body!!.pageNumber).isGreaterThan(0)
        assertThat(response.body!!.pageSize).isGreaterThan(0)
        assertThat(response.body!!.totalElements).isGreaterThan(0)
        assertThat(response.body!!.totalPages).isGreaterThan(0)
    }

    @Test
    fun listOwnersSupportsPagination() {
        val page1: ResponseEntity<PagedResultDto<Any>> = restTemplate.getForEntity("/api/v1/owners?page=1&size=2", PagedResultDto::class.java) as ResponseEntity<PagedResultDto<Any>>
        assertThat(page1.body).isNotNull()
        assertThat(page1.body!!.content.size).isLessThanOrEqualTo(2)

        val page2: ResponseEntity<PagedResultDto<Any>> = restTemplate.getForEntity("/api/v1/owners?page=2&size=2", PagedResultDto::class.java) as ResponseEntity<PagedResultDto<Any>>
        assertThat(page2.body).isNotNull()
    }

    @Test
    fun getOwnerReturnsCompleteDto() {
        val response: ResponseEntity<OwnerDto> = restTemplate.getForEntity("/api/v1/owners/1", OwnerDto::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        val owner = response.body
        assertThat(owner).isNotNull()
        assertThat(owner!!.id).isNotNull()
        assertThat(owner.firstName).isNotNull()
        assertThat(owner.lastName).isNotNull()
        assertThat(owner.address).isNotNull()
        assertThat(owner.city).isNotNull()
        assertThat(owner.telephone).isNotNull()
    }

    @Test
    fun createOwnerReturnsCreatedWithLocation() {
        val dto = OwnerCreateDto("Jane", "Doe", "456 Oak Ave", "Portland", "5035551234", "jane@doe.com")
        val response: ResponseEntity<OwnerDto> = restTemplate.postForEntity("/api/v1/owners", dto, OwnerDto::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.CREATED)
        assertThat(response.body).isNotNull()
        assertThat(response.body!!.id).isNotNull()
        assertThat(response.body!!.firstName).isEqualTo("Jane")
        assertThat(response.body!!.lastName).isEqualTo("Doe")
    }

    @Test
    fun createOwnerValidatesAllFields() {
        val blankDto = OwnerCreateDto("", "", "", "", "short", "invalid")
        val response: ResponseEntity<String> = restTemplate.postForEntity("/api/v1/owners", blankDto, String::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @Test
    fun createOwnerWithEmail() {
        val dto = OwnerCreateDto("Jane", "Doe", "456 Oak Ave", "Portland", "5035551234", "jane@example.com")
        val response: ResponseEntity<OwnerDto> = restTemplate.postForEntity("/api/v1/owners", dto, OwnerDto::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.CREATED)
        assertThat(response.body).isNotNull()
        assertThat(response.body!!.email).isEqualTo("jane@example.com")
    }

    @Test
    fun createOwnerWithoutEmail() {
        val dto = OwnerCreateDto("John", "Smith", "789 Pine St", "Seattle", "2065551234", null)
        val response: ResponseEntity<OwnerDto> = restTemplate.postForEntity("/api/v1/owners", dto, OwnerDto::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.CREATED)
        assertThat(response.body).isNotNull()
        assertThat(response.body!!.email).isNull()
    }

    @Test
    fun updateOwnerModifiesExistingOwner() {
        val dto = OwnerUpdateDto("UpdatedName", "Franklin", "110 W. Liberty St.", "Madison", "6085551023", "updated@email.com")
        val response: ResponseEntity<OwnerDto> = restTemplate.exchange(
            "/api/v1/owners/1", HttpMethod.PUT, HttpEntity(dto), OwnerDto::class.java
        )
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body!!.firstName).isEqualTo("UpdatedName")
    }

    @Test
    fun deleteOwnerRemovesOwner() {
        val dto = OwnerCreateDto("Delete", "Owner", "789 Pine", "Town", "9998887776", null)
        val createResponse: ResponseEntity<OwnerDto> = restTemplate.postForEntity("/api/v1/owners", dto, OwnerDto::class.java)
        val id = createResponse.body!!.id

        val deleteResponse: ResponseEntity<Void> = restTemplate.exchange(
            "/api/v1/owners/$id", HttpMethod.DELETE, null, Void::class.java
        )
        assertThat(deleteResponse.statusCode).isEqualTo(HttpStatus.NO_CONTENT)

        val getResponse: ResponseEntity<String> = restTemplate.getForEntity("/api/v1/owners/$id", String::class.java)
        assertThat(getResponse.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }

}